### PR TITLE
Report if basis functions are cellwise constant

### PR DESCRIPTION
**Pull request by miklos1**
This change makes FInAT report if a basis function is cellwise constant, so TSFC can optimise evaluation. This is achieved through an **API change**:

`basis_evaluation` no longer returns point structure as shape, but rather as free indices. If the value of the basis functions does not depend on some point indices, then those indices simply won't appear in the free indices of the return expression.

### CODE REVIEW

File: finat/fiat_elements.py [hunk 1]
```diff
@@ -32,36 +31,46 @@ def basis_evaluation(self, ps, entity=None, derivative=0):
...
 
-        fiat_tab = self._fiat_element.tabulate(derivative, ps.points, entity)
+        if derivative < self._degree:
+            points = ps.points
+            point_indices = ps.indices
+        elif derivative == self._degree:
+            # Tabulate on cell centre
+            points = np.mean(self._cell.get_vertices(), axis=0, keepdims=True)
+            entity = (self._cell.get_dimension(), 0)
+            point_indices = ()  # no point indices used
```
Comment (miklos1) [inline finat/fiat_elements.py]: Evaluation at cell centre when the specific derivative is cellwise constant.

File: finat/fiat_elements.py [hunk 2]
```diff
@@ -32,36 +31,46 @@ def basis_evaluation(self, ps, entity=None, derivative=0):
...
+        else:
+            return gem.Zero(tuple(index.extent for index in i + vi + di))
```
Comment (miklos1) [inline finat/fiat_elements.py]: Return appropriately shaped zeros for high derivatives.

File: finat/fiat_elements.py [hunk 3]
```diff
@@ -32,36 +31,46 @@ def basis_evaluation(self, ps, entity=None, derivative=0):
...
-            tensor = gem.ListTensor(tensor)
+            tensor = gem.Indexed(gem.ListTensor(tensor), di)
         else:
-            tensor = gem.Literal(restore_shape(fiat_tab[(0,) * dim].transpose(tr), ps))
+            tensor = tensor[()]
 
-        return gem.ComponentTensor(gem.Indexed(tensor,
-                                               di + pi + i + vi),
-                                   pi + i + vi + di)
+        return gem.ComponentTensor(tensor, i + vi + di)
```
Comment (wence-) [inline finat/fiat_elements.py]: OK, removing point indices from the returned CT, so they're free.

File: finat/fiat_elements.py [hunk 4]
```diff
@@ -32,36 +31,46 @@ def basis_evaluation(self, ps, entity=None, derivative=0):
         :param entity: the cell entity on which to tabulate.
         :param derivative: the derivative to take of the basis functions.
         '''
-
```
Comment (miklos1) [inline finat/fiat_elements.py]: This function looks a bit ugly, but I'm not sure yet how to clean it up.

**Comment by miklos1** [comment]
Merging this now, because the next pull request is coming up.

**Pull request merged**